Bgee parsing and unit tests

calliem opened the pull request.

calliem committed:
Added testing parser and preliminary conversion outlineg

calliem committed:
Preliminary bgee expression translation

calliem committed:
Merge branch 'master' of github.com:calliem/phenoscape-kb-ingest into bgee

calliem committed:
Renamed files and readded structure

calliem committed:
Added preliminary Java translation

calliem committed:
Merge branch 'master' of github.com:phenoscape/phenoscape-kb-ingest into bgee

calliem committed:
Merge

calliem committed:
Preliminary scala results

calliem committed:
IRI prefix and number of expressions test

calliem committed:
Extended to work on mice & tropicalis. Start of manchester syntax unit test

calliem committed:
Functional syntax comparison. Issues with differing NamedIndividual uuid's

calliem committed:
Build configuration updates

calliem committed:
Formatting

calliem committed:
Updated scala calls in Java

calliem committed:
Testing - build workspace error

balhoff committed:
Refactored imports.

calliem committed:
Test additions

calliem committed:
Merge branch 'master' of github.com:calliem/phenoscape-kb-ingest into bgee

calliem committed:
Merge pull request #1 from phenoscape/import-fixes

Refactored imports.

calliem committed:
Preprocessed results in Unit Test. Rewrote count & accuracy tests

calliem committed:
Preprocessed results in Unit Test. Rewrote count & accuracy tests

calliem committed:
Reorganized package hierarchy

calliem committed:
Reorganization

calliem committed:
Updated build path

balhoff commented on src/main/scala/org/phenoscape/kb/ingest/bgee/BgeeExpressionToOWL.scala:
```diff
@@ -0,0 +1,63 @@
+      if (term.startsWith("UBERON") || term.startsWith("CL")) {
+        val structureType = Class(OBOUtil.iriForTermID(Option(StringUtils.stripToNull(items(2))).filter(_ != "\\").get))
+        axioms.add(structure Type structureType)
+      } else if (term.startsWith("ZFA")) {
```
I think ZFA terms should expand into OBO IDs the same way UBERON and CL do. Any reason to do this differently? The `http://zfin.org/` IDs are not ontology classes (usually genes or genotypes).

balhoff commented on src/main/scala/org/phenoscape/kb/ingest/bgee/BgeeExpressionToOWL.scala:
```diff
@@ -0,0 +1,63 @@
+      val structure = OntUtil.nextIndividual()
+      axioms.add(Declaration(structure))
+
+      val term = (Option(StringUtils.stripToNull(items(2))).filter(_ != "\\").get)
```
If you wrap a result with `Option()`, and then call `option.get`, you are just undoing the wrapping. `Option` gives you a typesafe way to mark values that may possibly be null and force you to deal with that possibility. Java APIs often return null and so it can be good practice to wrap with `Option()`. The way you're doing it here will result in a `NullPointerException` if the string is empty. You should pretty much never call `get` on an option.

http://danielwestheide.com/blog/2012/12/19/the-neophytes-guide-to-scala-part-5-the-option-type.html

balhoff replied:
Oh, I see that I myself introduced this pattern in `ZFINExpressionToOWL`... :-D

ZFIN started putting backslashes in place of empty values and I wanted to make sure they were treated as empty (and I am okay with failing with NullPointerException because those fields are not supposed to be empty). So mostly ignore the previous comment, but perhaps there is no need for checking for backslashes with these data files? And no need to put the Option in.

balhoff commented:
@calliem I think this looks pretty good! Just see the couple of comments I put in the diffs. Instead of `Option(StringUtils.stripToNull(items(2))).filter(_ != "\\").get` I think you can pretty much just do `StringUtils.stripToNull(items(2))`. Bgee don't put backslashes in for empty values, do they? Make sure you check for null, or use Options and check for None, if it is okay for a value to be empty.

calliem committed:
Fixed term logic and definition

calliem commented:
@balhoff Thanks for the suggestions! I've changed how terms is defined and added ZFA to the conditional. If the results do happen to be null, what should be added to the axiom set in the Bgee case? 

I noticed in ZFINExpressionToOWL that if the substructureID was none, you added 

> ```
>       val structureType = Class(OBOUtil.iriForTermID(superStructureID))
>       axioms.add(structure Type structureType)
> ```

but it seems like ZFIN has a substructure in addition to a superstructure, while in Bgee we're looking at a singular anatomical entity ID.

I'm currently planning on simply skipping over these unspecified lines, but is there some kind of exception that should be built in if this case were to happen or is this something that shouldn't happen given the structure of the Bgee file?

balhoff commented:
It seems like it would be an error for there not to be a structure stated. So you can let it crash with a NullPointerException, but first maybe at least print an error message to the console so it is easy track down what went wrong.

calliem committed:
Fixed 'empty response' issue

calliem committed:
Added error checking for null structure values

calliem commented:
I added on a try-catch statement to check for it, although let me know if you had something else in mind for it.

calliem committed:
Ignore lines with 'ZFA'

hlapp commented on src/main/scala/org/phenoscape/kb/ingest/bgee/BgeeExpressionToOWL.scala:
```diff
@@ -25,7 +25,8 @@ object BgeeExpressionToOWL {
     val items = expressionLine.split("\t", -1)
-    if (items.length == 1 || items(6).startsWith("absent") || items(6).startsWith("Expression")) {
+    
+    if (items.length == 1 || items(6).startsWith("absent") || items(6).startsWith("Expression") || items(2).startsWith("ZFA")) {
```
@calliem could you add a comment here that we are skipping ZFA terms because these are all embryonic or otherwise pre-adult developmental stages that wouldn't make sense with the data the KB is otherwise populated with? Otherwise this reason may be lost when someone looks at it in the future.

hlapp commented on src/main/scala/org/phenoscape/kb/ingest/bgee/BgeeExpressionToOWL.scala:
```diff
@@ -0,0 +1,78 @@
+  def translate(expressionLine: String): Set[OWLAxiom] = {
+    val items = expressionLine.split("\t", -1)
+    
+    if (items.length == 1 || items(6).startsWith("absent") || items(6).startsWith("Expression") || items(2).startsWith("ZFA")) {
```
@calliem could you add a comment here that we are skipping ZFA terms because these are all embryonic or otherwise pre-adult developmental stages that wouldn't make sense with the data the KB is otherwise populated with? Otherwise this reason may be lost when someone looks at it in the future.

balhoff commented:
@calliem I think that is fine. You could even just check for null and print a message rather than catch the exception, since it is okay for this to crash (we would want to notice). But I think this is good, thanks!

balhoff commented:
I will just wait for @calliem to add the comment suggested by Hilmar and then I'll merge.

calliem commented:
@balhoff I added the comment but mistakenly pushed the change to master so it looks like the branch was merged automatically. Feel free to revert it if there are other issues remaining.

balhoff commented:
@calliem could you please open a pull request on the phenoscape repository? Then I'll merge in your code.

hlapp commented:
@balhoff this turned out to be little less straightforward that I would have thought, so I went ahead and manually put the merge on a PR, see phenoscape/phenoscape-kb-ingest#10. Can you review and comment there?
